packages/flame/lib/src/geometry/shape.dart

```diff
@@ -9,34 +10,125 @@ import 'shape_intersections.dart' as intersection_system;
+
+  /// The position of your shape in relation to its size from (-1,-1) to (1,1)
+  Vector2 relativeOffset = Vector2.zero();
+
+  /// The [relativeOffset] converted to a length vector
+  Vector2 get relativePosition => (size / 2)..multiply(relativeOffset);
+
+  /// The angle of the parent that has to be taken into consideration for some
+  /// applications of [Shape], for example [HitboxShape]
+  double parentAngle;
+
+  /// The center position of the shape within itself, without rotation
+  Vector2 get localCenter {
+    final stateValues = [
+      size,
+      relativeOffset,
+      offsetPosition,
+    ];
+    if (!_localCenterCache.isCacheValid(stateValues)) {
+      final center = (size / 2)..add(relativePosition)..add(offsetPosition);
+      _localCenterCache.updateCache(
+        center,
+        stateValues.map((e) => e.clone()).toList(growable: false),
+      );
+    }
+    return _localCenterCache.value!;
+  }
 
-  Vector2? _anchorPosition;
-  Vector2 get anchorPosition => _anchorPosition ?? position;
-  set anchorPosition(Vector2 position) => _anchorPosition = position;
+  /// The shape's absolute center with rotation taken into account
+  Vector2 get absoluteCenter {
+    final stateValues = [
+      position,
+      offsetPosition,
+      relativeOffset,
+      angle,
+      parentAngle,
+    ];
+    if (!_absoluteCenterCache.isCacheValid(stateValues)) {
+      /// The center of the shape, before any rotation
+      final center = position + offsetPosition;
+      if (!relativeOffset.isZero()) {
+        center.add(relativePosition);
+      }
+      if (angle != 0 || parentAngle != 0) {
+        center.rotate(parentAngle + angle, center: position);
+      }
+      _absoluteCenterCache.updateCache(center, [
+        position.clone(),
+        offsetPosition.clone(),
+        relativeOffset.clone(),
+        angle,
+        parentAngle,
+      ]);
+    }
+    return _absoluteCenterCache.value!;
+  }
+
+  Vector2 get renderCenter {
+    final zoom = isTranslated ? 1.0 : (camera?.zoom ?? 1);
+    return (isTranslated ? localCenter : absoluteCenter) / zoom;
+  }
+
+  /// Whether the canvas has been translated or not when it comes to rendering
+  /// the shape
+  final bool isTranslated = false;
+
+  /// Needed if you want your shapes to align with the zoom level set in the
+  /// camera
+  final Camera? camera;
 
   Shape({
     Vector2? position,
-    this.size,
+    Vector2? size,
     this.angle = 0,
-  }) : position = position ?? Vector2.zero();
+    this.parentAngle = 0,
+    this.camera,
+  })  : position = position ?? Vector2.zero(),
+        size = size ?? Vector2.zero();
 
   /// Whether the point [p] is within the shapes boundaries or not
   bool containsPoint(Vector2 p);
 
-  void render(Canvas c, Paint paint);
+  void render(Canvas canvas, Paint paint) {
+    canvas.save();
+    if (!isTranslated && camera != null) {
+      camera!.apply(canvas);
+    }
+    renderShape(canvas, paint);
+    canvas.restore();
+  }
```

Comment:
I'm not so sure about this having a camera. the way I see it:

shape -> camera agnostic, just renders
shapeComponent -> also camera agnostic, the base game does all camera transform
base game -> has the camera, does camera transforms
game -> no built in camera. you can add a camera and wrap render calls yourself

that is very high level idea though, lmk if it makes sense for this particular case.

Reply:
Then we would need to make `BaseGame` shape-aware then, or make a mixin for it?

Reply:
You're right, I created a `ShapeComponent`, now the code is much cleaner.